Call Transfer demo

DominicStewart opened the pull request:
This PR does several things:
1. Updates bot_runner to handle passing through json to the bot file. We have to handle dial in webhooks and dialing out for voicemail detection. 
2. Moves some of the code from bot_runner out into a separate file to make adding new examples in the future easier
3. Adds OpenAI Call Transfer examples
4. Removes the OpenAI Voicemail detection example. We want to make it easier to find the code you need. The readme explains what files to reference to understand how to adjust for a different LLM.
5. Moves some of the logic / state management for call transfer and voicemail detection to the call_connection_manager file
6. Updates voicemail detection to use this new logic

Previously one file handled both dial in and dial out use cases. But if a user is dialing in, there's no need to do voicemail detection.

The call transfer example is the example loaded when you dial in:

1. Human dials in to the bot
2. Human and bot has conversation
3. At some point, human escalates to an operator
4. When customer wants to escalate to an operator, bot will add the operator to the call
5. When operator joins the call, the bot gives a quick summary of the conversation
6. When customer is talking to an operator, bot just listens and does not respond (We prepend "Operator:" to the transcription frames when the operator speaks).
7. When the operator puts down the phone, the bot will let the customer know its still there and that the customer and the bot can continue their conversation
8. The bot now has the context on the conversation with the operator

If you dial out, even though the call transfer example could also work dial out (i.e. dial out to a user and then also escalate to an operator when the human escalates), we will tie the voicemail detection example to dial out.

The existing code for bot_twilio will still work as I haven't modified that.

vr000m commented:
Independent of the transfer implementation -- I see that the dialin_settings are received, do we send the complete dialin_settings to the bot? We should fix this because the dialin_settings contains to: and from: number, which most practical bots need access to understand who is calling.

vr000m commented on examples/phone-chatbot/bot_daily.py:
```diff
@@ -161,22 +334,60 @@ async def on_joined(transport, data):
+    @transport.event_handler("on_dialout_answered")
+    async def on_dialout_answered(transport, data):
+        logger.debug(f"Dial-out answered: {data}")
+        await transport.capture_participant_transcription(data["sessionId"])
```
shouldnt we keep track of all session_ids? i.e., if the original user was dialled out to and the escalation operator is a dialout as well?

DominicStewart replied:
For call transfer, we are only going to be showing how to dial in to the call, have a conversation and then escalate the call using dial out. Although yes, you could technically dial out to a user, and escalate to an operator. I've included code that handles customer / operator and bot session ids. But I haven't included that logic in this file.

vr000m commented:
for the transfer, maybe we should have a different bot file? Mainly to have a simple example for dialin-dialout, and then complex bot that has different variants depending on the type of transfer?

- simple dialout to operator, to bring the operator into the existing call between the bot and the original caller

Then there are variants of this experience, for example: the call needs to be transferred to a phone number, but the original user needs to navigate the IVR themselves.

DominicStewart commented:
@vr000m okay, i've tested these out. Should be okay. But would appreciate any input.

DominicStewart commented:
Still need to update the README file

vr000m commented on examples/phone-chatbot/README.md:
```diff
@@ -1,209 +1,415 @@
+			"callId": "callid-read-only-string",
+			"callDomain": "callDomain-read-only-string"
+		},
+		"dialout_settings": {
```
just making sure this is not an array but a single object

DominicStewart replied:
Ah, good one. It should be an array. I'll update the docs.

vr000m commented on examples/phone-chatbot/README.md:
````diff
@@ -1,209 +1,415 @@
-### Build the docker image
+   ```shell
+   mv fly.example.toml fly.toml
+   ```
````
Probably a link to pipecat cloud deployments -- quickstart and the starter-docker-images

DominicStewart replied:
Yeah makes sense. I just left that stuff in for now but plan to update with the Pipecat Cloud stuff

DominicStewart replied:
Lets discuss in a call what we want to include in this example folder RE docker images etc. Now we have Pipecat Cloud. I'm thinking we remove this fly.example.toml and the docker images from this folder. I haven't been keeping them up to date as I've updated the example.

DominicStewart replied:
For now i've updated the docs and removed the various docker images etc from the folder. I think this makes sense.

vr000m commented on examples/phone-chatbot/bot_runner.py:
```diff
@@ -29,218 +37,544 @@
+    errors = []
 
-app = FastAPI(lifespan=lifespan)
+    # Cannot have both dialin and dialout settings (unless in test mode)
```
is there a reason for this restriction?

DominicStewart replied:
> is there a reason for this restriction?

I think at the time I was thinking of how to detect the different examples. Probably not needed anymore. I'll adjust this.

vr000m commented on examples/phone-chatbot/bot_runner.py:
```diff
@@ -29,218 +37,544 @@
+    if has_call_transfer and has_voicemail:
+        errors.append(
+            "Cannot have both 'call_transfer' and 'voicemail_detection' in the same configuration"
+        )
```
Ok, this restriction makes sense. because the human user could leave a message or disconnect.

vr000m replied:
maybe make a table in readme.md which has checkboxes for which features are allowed in combination

DominicStewart replied:
Added a table to the readme file to illustrate which features are allowed in combination.

vr000m commented on examples/phone-chatbot/bot_runner.py:
```diff
@@ -29,218 +37,544 @@
+    """
     if not room_url:
-        # Create base properties with SIP settings
+        # Create base properties
```
we are setting these properties irrespective of dialin_settings is set?

vr000m commented on examples/phone-chatbot/bot_runner.py:
```diff
@@ -29,218 +37,544 @@
+        # Create base properties
         properties = DailyRoomProperties(
             sip=DailyRoomSipParams(
                 display_name="dialin-user", video=False, sip_mode="dial-in", num_endpoints=1
```
you may need to num_endpoints =2, if we do warm transfers between two daily rooms (more like future proof it).

vr000m commented on examples/phone-chatbot/bot_runner.py:
```diff
@@ -29,218 +37,544 @@
+            "to": data.get("To", ""),
+            "from": data.get("From", ""),
+            "callId": data.get("callId", data.get("CallSid", "")),
+            "callDomain": data.get("callDomain", ""),
```
we need to change these to snake_case somewhere, i.e., call_id and call_domain

markbackman commented:
@DominicStewart there's benefit in keeping a basic version of this demo that doesn't have bells and whistles. This will help from a learning process. You can see a version of this in our pstn_sip Pipecat Cloud image: https://github.com/daily-co/pipecat-cloud-images/tree/main/pipecat-starters/pstn_sip. If you jump directly to layering advanced topics, it can be difficult to understand how to do basic things, like make a call with Daily.

I would propose a daily_simple_bot.py and then whatever additional variants you think make sense to show.

DominicStewart commented:
> @DominicStewart there's benefit in keeping a basic version of this demo that doesn't have bells and whistles. This will help from a learning process. You can see a version of this in our pstn_sip Pipecat Cloud image: https://github.com/daily-co/pipecat-cloud-images/tree/main/pipecat-starters/pstn_sip. If you jump directly to layering advanced topics, it can be difficult to understand how to do basic things, like make a call with Daily.
> 
> I would propose a daily_simple_bot.py and then whatever additional variants you think make sense to show.

Ah yes, we will add that too. A simple dialin and a simple dialout bot

DominicStewart commented:
@markbackman I've added a simple dialin and dialout example. I'll run some more tests later but they seem to work fine. Refactored the bot runner file a bit to make it easier to add examples going forward. I'll add these examples to the README too.

markbackman commented:
> @markbackman I've added a simple dialin and dialout example. I'll run some more tests later but they seem to work fine. Refactored the bot runner file a bit to make it easier to add examples going forward. I'll add these examples to the README too.

Thanks @DominicStewart! One more question: is there a need to show examples for both OpenAI and Gemini? Could we simplify and just show one example? We added two examples for simple-chatbot due to the Gemini launch. If there's a need here, that's totally fine. Otherwise, maybe we just pick one LLM vendor to keep the examples streamlined.

DominicStewart commented:
Not necessarily, but for call transfer / voicemail detection it seems
things are different enough for Gemini and OpenAI that it made sense. I’m
happy to just keep Gemini for the voicemail detection one and OpenAI for
the rest. But they also exist now so not a big deal to keep them. Unless
you think it’s too much?

On Mon, 31 Mar 2025 at 20:29, Mark Backman ***@***.***> wrote:

> @markbackman <https://github.com/markbackman> I've added a simple dialin
> and dialout example. I'll run some more tests later but they seem to work
> fine. Refactored the bot runner file a bit to make it easier to add
> examples going forward. I'll add these examples to the README too.
>
> Thanks @DominicStewart <https://github.com/DominicStewart>! One more
> question: is there a need to show examples for both OpenAI and Gemini?
> Could we simplify and just show one example? We added two examples for
> simple-chatbot due to the Gemini launch. If there's a need here, that's
> totally fine. Otherwise, maybe we just pick one LLM vendor to keep the
> examples streamlined.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pipecat-ai/pipecat/pull/1348#issuecomment-2765947438>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK5KNFRHY5255TTPWR664YD2XERKJAVCNFSM6AAAAABYUSNJNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVHE2DONBTHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: markbackman]*markbackman* left a comment (pipecat-ai/pipecat#1348)
> <https://github.com/pipecat-ai/pipecat/pull/1348#issuecomment-2765947438>
>
> @markbackman <https://github.com/markbackman> I've added a simple dialin
> and dialout example. I'll run some more tests later but they seem to work
> fine. Refactored the bot runner file a bit to make it easier to add
> examples going forward. I'll add these examples to the README too.
>
> Thanks @DominicStewart <https://github.com/DominicStewart>! One more
> question: is there a need to show examples for both OpenAI and Gemini?
> Could we simplify and just show one example? We added two examples for
> simple-chatbot due to the Gemini launch. If there's a need here, that's
> totally fine. Otherwise, maybe we just pick one LLM vendor to keep the
> examples streamlined.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pipecat-ai/pipecat/pull/1348#issuecomment-2765947438>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK5KNFRHY5255TTPWR664YD2XERKJAVCNFSM6AAAAABYUSNJNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVHE2DONBTHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

markbackman commented:
> Not necessarily, but for call transfer / voicemail detection it seems things are different enough for Gemini and OpenAI that it made sense. I’m happy to just keep Gemini for the voicemail detection one and OpenAI for the rest. But they also exist now so not a big deal to keep them. Unless you think it’s too much?
> […](#)
> On Mon, 31 Mar 2025 at 20:29, Mark Backman ***@***.***> wrote: @markbackman <https://github.com/markbackman> I've added a simple dialin and dialout example. I'll run some more tests later but they seem to work fine. Refactored the bot runner file a bit to make it easier to add examples going forward. I'll add these examples to the README too. Thanks @DominicStewart <https://github.com/DominicStewart>! One more question: is there a need to show examples for both OpenAI and Gemini? Could we simplify and just show one example? We added two examples for simple-chatbot due to the Gemini launch. If there's a need here, that's totally fine. Otherwise, maybe we just pick one LLM vendor to keep the examples streamlined. — Reply to this email directly, view it on GitHub <[#1348 (comment)](https://github.com/pipecat-ai/pipecat/pull/1348#issuecomment-2765947438)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AK5KNFRHY5255TTPWR664YD2XERKJAVCNFSM6AAAAABYUSNJNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVHE2DONBTHA> . You are receiving this because you were mentioned.Message ID: ***@***.***> [image: markbackman]*markbackman* left a comment ([pipecat-ai/pipecat#1348](https://github.com/pipecat-ai/pipecat/pull/1348)) <[#1348 (comment)](https://github.com/pipecat-ai/pipecat/pull/1348#issuecomment-2765947438)> @markbackman <https://github.com/markbackman> I've added a simple dialin and dialout example. I'll run some more tests later but they seem to work fine. Refactored the bot runner file a bit to make it easier to add examples going forward. I'll add these examples to the README too. Thanks @DominicStewart <https://github.com/DominicStewart>! One more question: is there a need to show examples for both OpenAI and Gemini? Could we simplify and just show one example? We added two examples for simple-chatbot due to the Gemini launch. If there's a need here, that's totally fine. Otherwise, maybe we just pick one LLM vendor to keep the examples streamlined. — Reply to this email directly, view it on GitHub <[#1348 (comment)](https://github.com/pipecat-ai/pipecat/pull/1348#issuecomment-2765947438)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AK5KNFRHY5255TTPWR664YD2XERKJAVCNFSM6AAAAABYUSNJNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRVHE2DONBTHA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I'm mainly thinking about ease of discovery for the future. We link out to demos a lot in chatting with developers. Having an example directory that's easy to find files is a priority, so yes, let's keep it simple. It's also less to maintain going forward.

markbackman commented:
> Makes sense. How do you feel about: Just Gemini for VM detection. OpenAI for call transfer / simple examples

Looks good to me!

DominicStewart commented:
Rebased this file. Lots of changes. I think its good though. I've tested all the examples and it works well. Would love to get this merged in so I can move on.

jamsea commented on examples/phone-chatbot/requirements.txt:
```diff
@@ -1,4 +1,4 @@
-pipecat-ai[daily,elevenlabs,openai,silero]
+pipecat-ai[daily,cartesia,openai,google,silero]
 fastapi
```
@DominicStewart we should pin this version

vr000m reviewed (approved):
LGTM, we can merge and then make more changes. I like that we have several different bot files but a common runner
